constrain-lib/src/main/java/org/gradle/dependency/constrain/lib/serialize/ConstraintsJsonReader.java

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.dependency.constrain.lib.serialize;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.StreamReadFeature;
+import com.fasterxml.jackson.core.util.DefaultIndenter;
+import com.fasterxml.jackson.core.util.DefaultPrettyPrinter;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import com.fasterxml.jackson.module.paramnames.ParameterNamesModule;
+import com.github.difflib.DiffUtils;
+import com.github.difflib.UnifiedDiffUtils;
+import com.github.difflib.patch.Patch;
+import com.networknt.schema.JsonSchema;
+import com.networknt.schema.JsonSchemaException;
+import com.networknt.schema.JsonSchemaFactory;
+import com.networknt.schema.SpecVersion;
+import com.networknt.schema.ValidationMessage;
+import org.gradle.dependency.constrain.lib.DependencyConstrainException;
+import org.gradle.dependency.constrain.lib.model.LoadedConstraint;
+import org.gradle.dependency.constrain.lib.model.LoadedConstraints;
+
+import javax.annotation.Nullable;
+import javax.annotation.WillClose;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.gradle.dependency.constrain.lib.serialize.ConstrainFileLoader.DEPENDENCY_CONSTRAINTS_FILE_PATH_RELATIVE;
+
+/**
+ * Loads the {@link LoadedConstraints} from an input stream of JSON.
+ */
+public class ConstraintsJsonReader {
+    private static final String CONSTRAINTS_SPECIFICATION = "schema/dependency-constraints-schema.json";
+
+    /**
+     * Parses the JSON from the given input stream and returns the {@link LoadedConstraints}.
+     */
+    public static LoadedConstraints readFromJson(@WillClose InputStream in) {
+        try {
+            return doReadFromJson(in);
+        } catch (Exception ex) {
```

Comment:
`doReadFromJson` doesn't throw any checked exceptions so this could be
```suggestion
        } catch (DependencyConstrainException ex) {
            throw ex;
        } catch (RuntimeException ex) {
```
rethrow `DependencyConstrainException` without wrapping them again.

Reply:
This matches the existing pattern used by `DependencyVerificationsXmlReader`. All exceptions are wrapped here, they are however unwrapped and thrown away higher in the stack to add a longer message which includes the file path of the file that failed to load.

https://github.com/gradle/gradle/blob/ae3758aa7698196c03e7290116ea4b13388986c0/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/verification/serializer/DependencyVerificationsXmlReader.java#L77-L94

Reply:
I don't see any value in following that pattern, unless that code discards the actual exception and only looks at the `cause`, which would make adding a redundant layer important.